### PR TITLE
bpo-45885: Specialize COMPARE_OP

### PR DESCRIPTION
**Pull request by sweeneyde**
https://bugs.python.org/issue45885

### CODE REVIEW

**Comment by sweeneyde** [comment]
Some specialization statistics are here: https://gist.github.com/sweeneyde/49cc3a9d074d56cf095cb0a42d13d7a4

This should eliminate a lot of the type lookup/indirection in PyObject_RichCompare.

I don't have pyperformance numbers yet.

**Comment by sweeneyde** [comment]
Hm. At least on my (somewhat unstable) machine (wsl ubuntu gcc with pgo), numbers were disappointing:

Slower (16):
- pickle: 11.6 us +- 1.5 us -> 12.5 us +- 0.6 us: 1.08x slower
- unpickle: 16.4 us +- 1.2 us -> 17.2 us +- 0.5 us: 1.05x slower
- spectral_norm: 118 ms +- 2 ms -> 123 ms +- 2 ms: 1.05x slower
- regex_dna: 207 ms +- 1 ms -> 215 ms +- 3 ms: 1.04x slower
- scimark_fft: 377 ms +- 10 ms -> 391 ms +- 9 ms: 1.04x slower
- regex_effbot: 3.19 ms +- 0.05 ms -> 3.30 ms +- 0.06 ms: 1.04x slower
- scimark_sparse_mat_mult: 5.64 ms +- 0.10 ms -> 5.76 ms +- 0.26 ms: 1.02x slower
- mako: 12.4 ms +- 0.1 ms -> 12.6 ms +- 0.1 ms: 1.02x slower
- regex_compile: 157 ms +- 4 ms -> 159 ms +- 2 ms: 1.01x slower
- pickle_pure_python: 386 us +- 18 us -> 391 us +- 5 us: 1.01x slower
- chaos: 85.8 ms +- 1.2 ms -> 86.9 ms +- 2.6 ms: 1.01x slower
- deltablue: 4.97 ms +- 0.10 ms -> 5.04 ms +- 0.14 ms: 1.01x slower
- 2to3: 284 ms +- 3 ms -> 287 ms +- 6 ms: 1.01x slower
- logging_silent: 118 ns +- 2 ns -> 119 ns +- 1 ns: 1.01x slower
- python_startup: 8.05 ms +- 0.06 ms -> 8.09 ms +- 0.07 ms: 1.00x slower
- python_startup_no_site: 6.06 ms +- 0.09 ms -> 6.08 ms +- 0.04 ms: 1.00x slower

Faster (13):
- scimark_sor: 157 ms +- 2 ms -> 151 ms +- 4 ms: 1.04x faster
- regex_v8: 24.8 ms +- 0.2 ms -> 23.9 ms +- 0.3 ms: 1.04x faster
- unpack_sequence: 42.1 ns +- 0.9 ns -> 40.7 ns +- 0.9 ns: 1.03x faster
- logging_format: 8.68 us +- 0.57 us -> 8.42 us +- 0.20 us: 1.03x faster
- pyflate: 551 ms +- 13 ms -> 539 ms +- 10 ms: 1.02x faster
- raytrace: 385 ms +- 8 ms -> 378 ms +- 13 ms: 1.02x faster
- scimark_monte_carlo: 78.5 ms +- 1.0 ms -> 77.5 ms +- 1.6 ms: 1.01x faster
- scimark_lu: 119 ms +- 4 ms -> 118 ms +- 2 ms: 1.01x faster
- chameleon: 7.64 ms +- 0.13 ms -> 7.56 ms +- 0.10 ms: 1.01x faster
- logging_simple: 7.57 us +- 0.15 us -> 7.50 us +- 0.11 us: 1.01x faster
- float: 84.7 ms +- 1.3 ms -> 84.0 ms +- 1.1 ms: 1.01x faster
- xml_etree_process: 62.2 ms +- 0.7 ms -> 61.9 ms +- 0.8 ms: 1.01x faster
- dulwich_log: 83.1 ms +- 1.0 ms -> 82.6 ms +- 1.2 ms: 1.01x faster

Benchmark hidden because not significant (26): crypto_pyaes, django_template, fannkuch, go, hexiom, json_dumps, json_loads, meteor_contest, nbody, nqueens, pathlib, pickle_dict, pickle_list, pidigits, richards, sympy_expand, sympy_integrate, sympy_sum, sympy_str, telco, tornado_http, unpickle_list, unpickle_pure_python, xml_etree_parse, xml_etree_iterparse, xml_etree_generate

Geometric mean: 1.00x slower

**Comment by sweeneyde** [comment]
About 1.2x faster on microbenchmarks

```python
from pyperf import Runner
runner = Runner()

runner.timeit(
    "float_loop",
    setup="arr = [float(x) for x in range(5_000_000)]",
    stmt="for x in arr:\n"
         "    if x > 1e6:\n"
         "        break\n",
)

runner.timeit(
    "int_loop",
    setup="arr = list(range(5_000_000))",
    stmt="for i in arr:\n"
         "    if i == 6_000_000:\n"
         "        break\n",
)

runner.timeit(
    "str_loop",
    setup="arr = ['Py'] * 5_000_000",
    stmt="for word in arr:\n"
         "    if word == 'Python':\n"
         "        break\n",
)

"""
float_loop: Mean +- std dev: [compare_before] 17.7 ms +- 0.3 ms -> [compare_after] 14.3 ms +- 0.1 ms: 1.24x faster
int_loop: Mean +- std dev: [compare_before] 87.7 ms +- 1.4 ms -> [compare_after] 74.7 ms +- 1.2 ms: 1.17x faster
str_loop: Mean +- std dev: [compare_before] 85.3 ms +- 4.5 ms -> [compare_after] 70.2 ms +- 2.6 ms: 1.21x faster

Geometric mean: 1.21x faster
"""
```

**Comment by markshannon** [comment]
You microbenchmark is only testing `<`. What happens if you have a mix of comparisons?

File: Python/ceval.c [hunk 1]
```diff
@@ -3684,6 +3686,104 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
+        }
+
+        TARGET(COMPARE_OP_STR) {
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            PyObject *right = TOP();
+            PyObject *left = SECOND();
+            DEOPT_IF(!PyUnicode_CheckExact(left), COMPARE_OP);
+            DEOPT_IF(!PyUnicode_CheckExact(right), COMPARE_OP);
+            DEOPT_IF(!PyUnicode_IS_READY(left), COMPARE_OP);
```
Comment (markshannon) [inline Python/ceval.c]: Is this necessary? If `left == right` then they are equal regardless of whether the string is ready or not.

File: Python/ceval.c [hunk 2]
```diff
@@ -3684,6 +3686,104 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
+            DEOPT_IF(!PyUnicode_IS_READY(right), COMPARE_OP);
+            STAT_INC(COMPARE_OP, hit);
+            assert(cache0->original_oparg == Py_EQ || cache0->original_oparg == Py_NE);
+            int cmp = Py_Is(left, right) || _PyUnicode_EQ(left, right);
```
Comment (markshannon) [inline Python/ceval.c]: I don't think `_PyUnicode_EQ` is correct here. The docs say that a hash equality check should already have been performed.

File: Python/specialize.c [hunk 3]
```diff
@@ -143,6 +144,11 @@ _Py_GetSpecializationStats(void) {
 static void
 print_stats(FILE *out, SpecializationStats *stats, const char *name)
 {
+    long total = (stats->specialization_failure + stats->specialization_success
```
Comment (markshannon) [inline Python/specialize.c]: Please remove this, it breaks the summarizer script.

File: Python/specialize.c [hunk 4]
```diff
@@ -1539,3 +1551,48 @@ _Py_Specialize_BinaryOp(PyObject *lhs, PyObject *rhs, _Py_CODEUNIT *instr,
...
+        *instr = _Py_MAKECODEUNIT(COMPARE_OP_INT, _Py_OPARG(*instr));
+        goto success;
+    }
+    else if (PyUnicode_CheckExact(lhs)) {
+        int op = adaptive->original_oparg;
+        if (op != Py_EQ && op != Py_NE) {
+            SPECIALIZATION_FAIL(COMPARE_OP, SPEC_FAIL_STRING_COMPARE);
+            goto failure;
+        }
+        else if (!PyUnicode_IS_READY(lhs) || !PyUnicode_IS_READY(rhs)) {
```
Comment (markshannon) [inline Python/specialize.c]: There is no point in this check. The instruction needs to check for readiness regardless.

**Comment by markshannon** [comment]
There are a lot of branches in the int and float specializations, which is not good for performance.
https://github.com/python/cpython/blob/main/Python/adaptive.md#implementation-of-specialized-instructions

Because neither comparison can fail (ints have a total ordering, and floats have a total ordering plus a special not-equal-if-either-is-a-nan result) we can avoid the branching by masking the result of a comparison with a code for the operator.
E.g.  `res = (float_compare() & OPERATOR_MASK) ? Py_True : Py_False`

Also see my comment on the issue about avoiding the need to push and pop objects.

File: Python/ceval.c (hunk 1 above)
Comment (sweeneyde) [reply Python/ceval.c]: I expect that almost all strings are ready. I suppose the alternative is
```
            DEOPT_IF(!PyUnicode_CheckExact(left), COMPARE_OP);
            DEOPT_IF(!PyUnicode_CheckExact(right), COMPARE_OP);
            int res = 1;
            if (left != right) {
                DEOPT_IF(!PyUnicode_IS_READY(left), COMPARE_OP);
                DEOPT_IF(!PyUnicode_IS_READY(left), COMPARE_OP);
                res = comare_the_strings_assuming_theyre_ready(left, right);
            }
```
is that what you mean?

File: Python/ceval.c (hunk 2 above)
Comment (sweeneyde) [reply Python/ceval.c]: Using PyUnicode_RichCompare would call PyUnicode_READY on its arguments, which should almost always just check that PyUnicode_IS_READY. PyUnicode_RichCompare then passes to unicode_compare_eq, while _PyUnicode_EQ passes to unicode_eq in stringlib/eq.h. The implementations of unicode_eq and unicode_compare_eq are essentially the same, but _PyUnicode_EQ was already exposed, so I went with that.

Is the concern the different branching characteristics of this versus the very-likely-equal hash table case? Other than that, the function seems to do exactly what you would expect.

File: Python/specialize.c (hunk 4 above)
Comment (sweeneyde) [reply Python/specialize.c]: If a bytecode is going to fail the readiness check during specialization, isn't it likely to fail the readiness check afterwards? My worry was that without the check, the following code would needlessly re-optimize and de-optimize repeatedly:

```
for s in unready_strings_from_c_extension():
    if s == "Python":
        break
```

**Comment by sweeneyde** [comment]
> res = (float_compare() & OPERATOR_MASK) ? Py_True : Py_False

Is float_compare just `dright-dleft`, with `OPERATOR_MASK` checking the sign bit of the double? Is that safe across platforms?

For ints, I suppose it's:
```
Py_NE => cmp & 0xFFFF...FF
Py_LT => cmp & 0x8000...00
Py_LE => (cmp - 1) & 0x8000...00
Py_EQ => !cmp
Py_GT => (-cmp) & 0x8000...00
Py_GE => (-(cmp - 1)) & 0x8000...00
```
How can those get unified? An &-mask followed by a ^-mask or something?

**Comment by sweeneyde** [comment]
Here's one idea for ints:

```
cmp = ...
// same as sign = (cmp > 0) - (cmp < 0)
int sign = (int)((cmp>>63) - ((-cmp)>>63));
int flags = 1 << (sign + 1);
int res = flags & mask;
// mask: 0b001 is negative, 0b010 is zero, 0b100 is positive, etc.
next_instruction()
if (res) JUMP_BY(oparg);
```

Then the mask can be chosen depending on whether the following instruction is pop_jump_if_true/false.

There would still be the branch depending on whether Py_SIZE(left) - Py_SIZE(right) == 0 and a loop over digits and a branch to check for Py_SIZE(left) < 0, but at least it would save the switch over oparg, the push/pop/incref/decref, and one DISPATCH().

Maybe the float equivalent would deopt if either arg is NaN and then do `int sign = (dleft > dright) - (dleft < dright)`.

**Comment by sweeneyde** [comment]
These three combined opcodes have [slightly worse hit percentages on some benchmarks](https://gist.github.com/sweeneyde/07488eb1286acfbe14ce659bec76ddd3).

**Comment by sweeneyde** [comment]
Newer Pyperformance numbers:

Slower (20):
- unpickle: 16.4 us +- 1.5 us -> 17.8 us +- 0.9 us: 1.08x slower
- pickle_list: 4.28 us +- 0.09 us -> 4.63 us +- 0.13 us: 1.08x slower
- scimark_sparse_mat_mult: 5.49 ms +- 0.21 ms -> 5.89 ms +- 0.16 ms: 1.07x slower
- pickle: 12.3 us +- 0.9 us -> 13.0 us +- 1.1 us: 1.06x slower
- regex_dna: 205 ms +- 3 ms -> 217 ms +- 5 ms: 1.06x slower
- meteor_contest: 108 ms +- 3 ms -> 112 ms +- 2 ms: 1.04x slower
- pickle_dict: 28.4 us +- 0.2 us -> 29.5 us +- 0.7 us: 1.04x slower
- python_startup: 8.00 ms +- 0.06 ms -> 8.25 ms +- 0.50 ms: 1.03x slower
- logging_silent: 114 ns +- 1 ns -> 118 ns +- 2 ns: 1.03x slower
- regex_effbot: 3.16 ms +- 0.02 ms -> 3.25 ms +- 0.03 ms: 1.03x slower
- nqueens: 95.2 ms +- 1.9 ms -> 97.7 ms +- 1.8 ms: 1.03x slower
- unpickle_list: 4.97 us +- 0.11 us -> 5.10 us +- 0.10 us: 1.03x slower
- nbody: 101 ms +- 5 ms -> 103 ms +- 3 ms: 1.03x slower
- python_startup_no_site: 6.01 ms +- 0.04 ms -> 6.16 ms +- 0.27 ms: 1.03x slower
- mako: 12.3 ms +- 0.1 ms -> 12.5 ms +- 0.1 ms: 1.02x slower
- logging_simple: 7.51 us +- 0.13 us -> 7.67 us +- 0.16 us: 1.02x slower
- spectral_norm: 114 ms +- 4 ms -> 116 ms +- 2 ms: 1.02x slower
- scimark_fft: 378 ms +- 7 ms -> 385 ms +- 12 ms: 1.02x slower
- tornado_http: 183 ms +- 7 ms -> 186 ms +- 6 ms: 1.02x slower
- pathlib: 20.6 ms +- 0.3 ms -> 20.8 ms +- 0.4 ms: 1.01x slower

Faster (17):
- scimark_sor: 153 ms +- 4 ms -> 142 ms +- 4 ms: 1.08x faster
- unpack_sequence: 43.0 ns +- 1.0 ns -> 41.0 ns +- 0.8 ns: 1.05x faster
- go: 166 ms +- 2 ms -> 159 ms +- 3 ms: 1.04x faster
- django_template: 44.4 ms +- 1.8 ms -> 42.6 ms +- 2.5 ms: 1.04x faster
- fannkuch: 433 ms +- 10 ms -> 419 ms +- 10 ms: 1.03x faster
- pidigits: 183 ms +- 1 ms -> 178 ms +- 1 ms: 1.03x faster
- scimark_monte_carlo: 78.5 ms +- 1.7 ms -> 76.5 ms +- 1.1 ms: 1.03x faster
- chameleon: 7.69 ms +- 0.13 ms -> 7.50 ms +- 0.10 ms: 1.03x faster
- telco: 6.96 ms +- 0.13 ms -> 6.83 ms +- 0.19 ms: 1.02x faster
- hexiom: 7.49 ms +- 0.13 ms -> 7.35 ms +- 0.18 ms: 1.02x faster
- pickle_pure_python: 394 us +- 6 us -> 387 us +- 9 us: 1.02x faster
- deltablue: 4.98 ms +- 0.10 ms -> 4.90 ms +- 0.14 ms: 1.02x faster
- scimark_lu: 118 ms +- 2 ms -> 116 ms +- 2 ms: 1.02x faster
- 2to3: 282 ms +- 4 ms -> 278 ms +- 4 ms: 1.01x faster
- float: 84.2 ms +- 1.2 ms -> 83.1 ms +- 1.6 ms: 1.01x faster
- unpickle_pure_python: 298 us +- 7 us -> 294 us +- 6 us: 1.01x faster
- pyflate: 550 ms +- 14 ms -> 545 ms +- 8 ms: 1.01x faster

Benchmark hidden because not significant (18): chaos, crypto_pyaes, dulwich_log, json_dumps, json_loads, logging_format, raytrace, regex_compile, regex_v8, richards, sympy_expand, sympy_integrate, sympy_sum, sympy_str, xml_etree_parse, xml_etree_iterparse, xml_etree_generate, xml_etree_process

Geometric mean: 1.00x slower

**Comment by sweeneyde** [comment]
Newest microbenchmarks: pretty good, even including more comparisons


```python
from pyperf import Runner
runner = Runner()

runner.timeit(
    "float_loop",
    setup="arr = [float(x) for x in range(5_000_000)]",
    stmt="for x in arr:\n"
         "    if x > 1e6 or x == 1e6 or x < 0.0:\n"
         "        break\n",
)

runner.timeit(
    "int_loop",
    setup="arr = list(range(5_000_000))",
    stmt="for i in arr:\n"
         "    if i > 6_000_000 or i == 6_000_000 or i < 0:\n"
         "        break\n",
)

runner.timeit(
    "str_loop",
    setup="arr = ['Py'] * 5_000_000",
    stmt="for word in arr:\n"
         "    if word == 'Python' or word != 'Py':\n"
         "        break\n",
)

"""
float_loop: Mean +- std dev: [micro_main] 41.5 ms +- 0.4 ms -> [micro_combined3] 23.7 ms +- 0.8 ms: 1.75x faster
int_loop: Mean +- std dev: [micro_main] 202 ms +- 2 ms -> [micro_combined3] 133 ms +- 1 ms: 1.52x faster
str_loop: Mean +- std dev: [micro_main] 140 ms +- 7 ms -> [micro_combined3] 88.1 ms +- 3.9 ms: 1.59x faster

Geometric mean: 1.62x faster
"""

```

**Comment by sweeneyde** [comment]
Another run pair of runs with the latest commit:

Slower (16):
- unpack_sequence: 41.3 ns +- 0.7 ns -> 44.2 ns +- 1.3 ns: 1.07x slower
- scimark_sparse_mat_mult: 5.28 ms +- 0.30 ms -> 5.57 ms +- 0.11 ms: 1.05x slower
- nbody: 92.6 ms +- 2.6 ms -> 96.8 ms +- 4.7 ms: 1.05x slower
- raytrace: 370 ms +- 13 ms -> 381 ms +- 11 ms: 1.03x slower
- unpickle_list: 4.98 us +- 0.11 us -> 5.09 us +- 0.14 us: 1.02x slower
- deltablue: 5.06 ms +- 0.11 ms -> 5.16 ms +- 0.12 ms: 1.02x slower
- django_template: 44.4 ms +- 1.8 ms -> 45.2 ms +- 1.8 ms: 1.02x slower
- pickle_dict: 27.9 us +- 0.2 us -> 28.4 us +- 0.2 us: 1.02x slower
- sympy_str: 352 ms +- 12 ms -> 357 ms +- 7 ms: 1.01x slower
- nqueens: 94.2 ms +- 2.4 ms -> 95.4 ms +- 1.9 ms: 1.01x slower
- sympy_expand: 590 ms +- 21 ms -> 598 ms +- 9 ms: 1.01x slower
- pathlib: 20.4 ms +- 0.4 ms -> 20.7 ms +- 0.3 ms: 1.01x slower
- go: 165 ms +- 3 ms -> 167 ms +- 3 ms: 1.01x slower
- regex_compile: 157 ms +- 2 ms -> 158 ms +- 3 ms: 1.01x slower
- json_dumps: 13.1 ms +- 0.3 ms -> 13.2 ms +- 0.3 ms: 1.01x slower
- python_startup_no_site: 6.08 ms +- 0.04 ms -> 6.12 ms +- 0.05 ms: 1.01x slower

Faster (15):
- scimark_sor: 153 ms +- 4 ms -> 143 ms +- 4 ms: 1.07x faster
- regex_effbot: 3.35 ms +- 0.03 ms -> 3.20 ms +- 0.02 ms: 1.05x faster
- json_loads: 28.8 us +- 2.0 us -> 27.4 us +- 1.7 us: 1.05x faster
- pickle_pure_python: 392 us +- 13 us -> 381 us +- 15 us: 1.03x faster
- pyflate: 554 ms +- 13 ms -> 544 ms +- 11 ms: 1.02x faster
- pidigits: 181 ms +- 1 ms -> 178 ms +- 1 ms: 1.02x faster
- hexiom: 7.55 ms +- 0.17 ms -> 7.43 ms +- 0.19 ms: 1.02x faster
- regex_dna: 218 ms +- 6 ms -> 215 ms +- 5 ms: 1.02x faster
- crypto_pyaes: 92.5 ms +- 2.6 ms -> 91.1 ms +- 2.2 ms: 1.02x faster
- dulwich_log: 83.4 ms +- 1.1 ms -> 82.3 ms +- 1.4 ms: 1.01x faster
- logging_silent: 116 ns +- 2 ns -> 115 ns +- 2 ns: 1.01x faster
- float: 84.6 ms +- 1.0 ms -> 83.7 ms +- 1.6 ms: 1.01x faster
- xml_etree_generate: 87.4 ms +- 1.0 ms -> 86.6 ms +- 1.2 ms: 1.01x faster
- sympy_integrate: 23.8 ms +- 0.3 ms -> 23.6 ms +- 0.3 ms: 1.01x faster
- python_startup: 8.16 ms +- 0.10 ms -> 8.14 ms +- 0.06 ms: 1.00x faster

Benchmark hidden because not significant (24): 2to3, chameleon, chaos, fannkuch, logging_format, logging_simple, mako, meteor_contest, pickle, pickle_list, regex_v8, richards, scimark_fft, scimark_lu, scimark_monte_carlo, spectral_norm, sympy_sum, telco, tornado_http, unpickle, unpickle_pure_python, xml_etree_parse, xml_etree_iterparse, xml_etree_process

Geometric mean: 1.00x slower

**Comment by sweeneyde** [comment]
Slower (19):
- pickle_dict: 27.9 us +- 0.2 us -> 29.9 us +- 0.2 us: 1.07x slower
- scimark_sparse_mat_mult: 5.28 ms +- 0.30 ms -> 5.63 ms +- 0.12 ms: 1.07x slower
- nbody: 92.6 ms +- 2.6 ms -> 98.6 ms +- 2.1 ms: 1.06x slower
- pickle_list: 4.28 us +- 0.11 us -> 4.51 us +- 0.11 us: 1.05x slower
- regex_v8: 23.5 ms +- 0.5 ms -> 24.5 ms +- 0.4 ms: 1.04x slower
- unpack_sequence: 41.3 ns +- 0.7 ns -> 42.9 ns +- 1.1 ns: 1.04x slower
- sympy_str: 352 ms +- 12 ms -> 364 ms +- 6 ms: 1.03x slower
- scimark_fft: 377 ms +- 10 ms -> 390 ms +- 7 ms: 1.03x slower
- django_template: 44.4 ms +- 1.8 ms -> 45.7 ms +- 1.6 ms: 1.03x slower
- json_dumps: 13.1 ms +- 0.3 ms -> 13.5 ms +- 0.3 ms: 1.03x slower
- raytrace: 370 ms +- 13 ms -> 380 ms +- 12 ms: 1.03x slower
- spectral_norm: 120 ms +- 3 ms -> 123 ms +- 2 ms: 1.03x slower
- sympy_expand: 590 ms +- 21 ms -> 603 ms +- 15 ms: 1.02x slower
- scimark_lu: 119 ms +- 5 ms -> 121 ms +- 4 ms: 1.02x slower
- sympy_sum: 193 ms +- 3 ms -> 196 ms +- 2 ms: 1.02x slower
- regex_effbot: 3.35 ms +- 0.03 ms -> 3.39 ms +- 0.04 ms: 1.01x slower
- nqueens: 94.2 ms +- 2.4 ms -> 95.1 ms +- 1.6 ms: 1.01x slower
- meteor_contest: 109 ms +- 2 ms -> 110 ms +- 2 ms: 1.01x slower
- pathlib: 20.4 ms +- 0.4 ms -> 20.6 ms +- 0.3 ms: 1.01x slower

Faster (20):
- scimark_sor: 153 ms +- 4 ms -> 136 ms +- 2 ms: 1.13x faster
- pyflate: 554 ms +- 13 ms -> 518 ms +- 13 ms: 1.07x faster
- scimark_monte_carlo: 78.8 ms +- 2.1 ms -> 75.3 ms +- 0.9 ms: 1.05x faster
- deltablue: 5.06 ms +- 0.11 ms -> 4.86 ms +- 0.12 ms: 1.04x faster
- fannkuch: 439 ms +- 6 ms -> 425 ms +- 19 ms: 1.03x faster
- unpickle_pure_python: 292 us +- 4 us -> 284 us +- 5 us: 1.03x faster
- json_loads: 28.8 us +- 2.0 us -> 28.1 us +- 0.8 us: 1.02x faster
- unpickle_list: 4.98 us +- 0.11 us -> 4.87 us +- 0.12 us: 1.02x faster
- mako: 12.6 ms +- 0.2 ms -> 12.4 ms +- 0.2 ms: 1.02x faster
- pickle_pure_python: 392 us +- 13 us -> 385 us +- 9 us: 1.02x faster
- 2to3: 288 ms +- 14 ms -> 284 ms +- 7 ms: 1.02x faster
- xml_etree_iterparse: 112 ms +- 3 ms -> 110 ms +- 2 ms: 1.02x faster
- crypto_pyaes: 92.5 ms +- 2.6 ms -> 91.1 ms +- 2.3 ms: 1.02x faster
- go: 165 ms +- 3 ms -> 162 ms +- 2 ms: 1.01x faster
- hexiom: 7.55 ms +- 0.17 ms -> 7.45 ms +- 0.14 ms: 1.01x faster
- logging_simple: 7.67 us +- 0.16 us -> 7.59 us +- 0.14 us: 1.01x faster
- dulwich_log: 83.4 ms +- 1.1 ms -> 82.6 ms +- 1.2 ms: 1.01x faster
- python_startup: 8.16 ms +- 0.10 ms -> 8.09 ms +- 0.17 ms: 1.01x faster
- logging_silent: 116 ns +- 2 ns -> 115 ns +- 2 ns: 1.01x faster
- chameleon: 7.51 ms +- 0.10 ms -> 7.46 ms +- 0.09 ms: 1.01x faster

Benchmark hidden because not significant (16): chaos, float, logging_format, pickle, pidigits, python_startup_no_site, regex_compile, regex_dna, richards, sympy_integrate, telco, tornado_http, unpickle, xml_etree_parse, xml_etree_generate, xml_etree_process

Geometric mean: 1.00x slower

**Comment by markshannon** [comment]
My [results](https://gist.github.com/markshannon/1d5ed186e520abbf980c7c82a39d6858) are less noisy and show a ~1% speedup.

Given that some of the largest improvements are from notoriously noisy benchmarks, 1% may be overstating the speedup but it does seem to be real.

File: Python/ceval.c (hunk 1 above)
Comment (markshannon) [reply Python/ceval.c]: Yes. Assuming that the identity shortcut is worthwhile, then it makes sense to shortcut as much work as possible.
You are also missing out of the quick inequality test using lengths `len(a) != len(b)` implies that `a != b`.

Maybe wrap `unicode_compare_eq` in an equality function that readies the strings (I'm surprised that no such thing exists) and use that?

```
int res = 1;
if (left != right) {
    res = _PyUnicode_Equal(left, right);
}

File: Python/ceval.c (hunk 2 above)
Comment (markshannon) [reply Python/ceval.c]: The docs say that a hash equality should already have been performed before using `_PyUnicode_EQ`, but you haven't done that.

So either the docs are wrong, or you code is wrong. From what you say, it sounds like the docs are wrong, and `_PyUnicode_EQ` can be used on any readied strings.

This is moot, of course, if you make the change discussed above.

File: Python/specialize.c (hunk 4 above)
Comment (markshannon) [reply Python/specialize.c]: Maybe ready the string here to avoid that?

File: Python/ceval.c [hunk 5]
```diff
@@ -3684,6 +3686,124 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
+                CHECK_EVAL_BREAKER();
+                NOTRACE_DISPATCH();
+            }
+        }
+
+        TARGET(COMPARE_OP_STR_JUMP) {
+            assert(cframe.use_tracing == 0);
+            // Combined: COMPARE_OP (str == str or str != str) + POP_JUMP_IF_(true/false)
+            SpecializedCacheEntry *caches = GET_CACHE();
+            int mask = caches[-1].compare.mask;
```
Comment (markshannon) [inline Python/ceval.c]: Could you change `mask` to a more descriptive name, say `branch_if_compares`?

File: Python/ceval.c [hunk 6]
```diff
@@ -3684,6 +3686,124 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
+            PyObject *right = TOP();
+            PyObject *left = SECOND();
+            DEOPT_IF(!PyFloat_CheckExact(left), COMPARE_OP);
+            DEOPT_IF(!PyFloat_CheckExact(right), COMPARE_OP);
+            double dleft = PyFloat_AS_DOUBLE(left);
+            double dright = PyFloat_AS_DOUBLE(right);
+            int sign = (dleft > dright) - (dleft < dright);
+            DEOPT_IF(isnan(dleft), COMPARE_OP);
```
Comment (markshannon) [inline Python/ceval.c:3720]: Potentially we could use the fact that nans are not equal to everything including themselves to add a fourth bit to the mask. Probably leave that for another PR as the maths starts getting a bit convoluted.

File: Include/internal/pycore_code.h [hunk 7]
```diff
@@ -42,6 +42,10 @@ typedef struct {
     uint16_t defaults_len;
 } _PyCallCache;
 
+typedef struct {
+    int mask;
+} _PyCompareCache;
```
Comment (markshannon) [inline Include/internal/pycore_code.h]: You only need a few bits, so this could easily fit into the `index` field of the `_PyAdaptiveEntry`

File: Python/specialize.c [hunk 8]
```diff
@@ -251,6 +254,7 @@ static uint8_t cache_requirements[256] = {
     [CALL_FUNCTION] = 2, /* _PyAdaptiveEntry and _PyObjectCache/_PyCallCache */
     [STORE_ATTR] = 2, /* _PyAdaptiveEntry and _PyAttrCache */
     [BINARY_OP] = 1,  // _PyAdaptiveEntry
+    [COMPARE_OP] = 2, /* _PyAdaptiveEntry and _PyCompareCache */
```
Comment (markshannon) [inline Python/specialize.c]: You only need 1 entry if you put the mask into the `_PyAdaptiveEntry`

File: Python/specialize.c [hunk 9]
```diff
@@ -1539,3 +1548,79 @@ _Py_Specialize_BinaryOp(PyObject *lhs, PyObject *rhs, _Py_CODEUNIT *instr,
...
+            SPECIALIZATION_FAIL(COMPARE_OP, SPEC_FAIL_STRING_UNREADY);
+            goto failure;
+        }
+        else {
+            *instr = _Py_MAKECODEUNIT(COMPARE_OP_STR_JUMP, _Py_OPARG(*instr));
+            cache1->mask = (mask & 2) == 0;
+            goto success;
+        }
+    }
+    SPECIALIZATION_FAIL(COMPARE_OP, SPEC_FAIL_OTHER);
```
Comment (markshannon) [inline Python/specialize.c:1612]: I'd like more information here.
Is this a builtin class, or a Python class? 
Does it override the comparison, or (in the case of `==/!==`) does it rely on the default identity comparison?

Doesn't need to be in this PR, though.

**Comment by markshannon** [comment]
Looks good in general. The approach seems sound and gives a speedup.

There a couple of things that need to be tightened up, and a couple of possible enhancements (for future PRs).

**Comment by markshannon** [comment]
[Updated results](https://gist.github.com/markshannon/55e2e3993eb7762f4d098b8c5f11168a) are good.

**Comment by markshannon** [comment]
:+1:

**Comment by joaoe** [comment]
Hi. I just stumbled upon this PR.
Now in Objects/unicodeobject.c there is both a `_PyUnicode_EQ` and `_PyUnicode_Equal`. Is that on purpouse ?
It looks a bit redundant, also because the former `_PyUnicode_Equal` uses `unicode_eq()` which does not do a pointer check before the more expensive `memcmp`. Meanwhile your new function uses the local function `unicode_compare_eq` which seems a duplicate of `unicode_eq`. Opportunity for some cleanup ?
